tests/foreman/virtwho/utils.py

```diff
@@ -0,0 +1,185 @@
+        }
+    else:
+        raise VirtWhoError(
+            '"{}" system type is not supported. Please use one of {}'
+            .format(system_type, ('satellite', 'guest'))
+        )
+
+
+def runcmd(cmd, system=None, timeout=None, port=22):
+    """Return the retcode and stdout.
+    :param str cmd: The command line will be executed in the target system.
+    :param dict system: the system account which ssh will connect to,
+        it will connect to the satellite host if the system is None.
+    :param int timeout: Time to wait for establish the connection.
+    :param int port: The server port to ssh connect tol
+    """
+    system = system or get_system('satellite')
+    hostname = system['hostname']
+    username = system['username']
+    password = system['password']
+    if ":" in hostname:
+        hostname, port = hostname.split(':')
+    result = ssh.command(
+            cmd, hostname=hostname, username=username,
+            password=password, timeout=timeout, port=int(port))
+    return result.return_code, '\n'.join(result.stdout)
+
+
+def register_system(system, activation_key=None, org='Default_Organization'):
+    """Return True if the system is registered to satellite successfully.
+    :param dict system: system account used by ssh to connect and register.
+    :param str activation_key: the activation key will be used to register.
+    :param str org: Which organization will be used to register.
+    :raises: VirtWhoError: If failed to register the system.
+    """
+    runcmd('subscription-manager unregister', system)
+    runcmd('subscription-manager clean', system)
+    runcmd('rpm -qa | grep katello-ca-consumer | xargs rpm -e |sort', system)
+    runcmd(
+        'rpm -ihv http://{}/pub/katello-ca-consumer-latest.noarch.rpm'
+        .format(settings.server.hostname), system
+    )
+    cmd = 'subscription-manager register --org={} '.format(org)
+    if activation_key is not None:
+        cmd += '--activationkey={}'.format(activation_key)
+    else:
+        cmd += '--username={} --password={}'.format(
+            settings.server.admin_username,
+            settings.server.admin_password)
+    ret, stdout = runcmd(cmd, system)
+    if ret == 0 or "system has been registered" in stdout:
+        return True
+    else:
+        raise VirtWhoError(
+            'Failed to register system: {}'.format(system)
+        )
+
+
+def virtwho_cleanup():
+    """Before running test cases, need to clean the environment.
+    Do the following:
+    1. stop virt-who service.
+    2. kill all the virt-who pid
+    3. clean rhsm.log message, make sure there is no old message exist.
+    4. clean all the configure files in /etc/virt-who.d/
+    """
+    runcmd("systemctl stop virt-who")
+    runcmd(
+        "ps -ef | grep virt-who -i | grep -v grep |"
+        "awk '{print $2}' | xargs -I {} kill -9 {}"
+    )
+    runcmd("rm -f /var/run/virt-who.pid")
+    runcmd("rm -rf /var/log/rhsm/*")
```

Comment:
Yes, I have to get the hypervisorId from /var/log/rhsm/rhsm.log, and then I can know which entry is the expected hypervisor in the satellite Content Hosts page, there is no other way to know the right hypervisorId, we only can get it from the rhsm.log file.

Meanwhile, I have to make sure there is no old json info or other unexpected json info exist in the rhsm.log, for example, if we run virt-who for vcenter, the json info will be recorded in rhsm.log, now we run virt-who for hyperv-v, there are two json info in the rhsm.log, this is why we need to delete the rhsm.log for each test case, always keep only on json info in the rhsm.log.
```
2019-07-18 02:11:44,093 [virtwho.destination_-1753402522345217122 DEBUG] MainProcess(21841):Thread-3 @subscriptionmanager.py:hypervisorCheckIn:207 - Host-to-guest mapping being sent to 'Default_Organization': {
    "hypervisors": [
        {
            "hypervisorId": {
                "hypervisorId": "bootp-73-131-226.rhts.eng.pek2.redhat.com"
            },
            "name": "bootp-73-131-226.rhts.eng.pek2.redhat.com",
            "guestIds": [
                {
                    "guestId": "564dd2ec-ed5b-658a-359f-2158dab0907f",
                    "state": 1,
                    "attributes": {
                        "active": 1,
                        "virtWhoType": "esx"
                    }
                }
            ],
            "facts": {
                "hypervisor.type": "VMware ESXi",
                "dmi.system.uuid": "9bff4d56-3b59-6355-ece2-e02094d27d11",
                "cpu.cpu_socket(s)": "1",
                "hypervisor.version": "6.7.0"
            }
        }
    ]
}

```

Reply:
> for example, if we run virt-who for vcenter, the json info will be recorded in rhsm.log, now we run virt-who for hyperv-v, there are two json info in the rhsm.log

You can tell which one is which based on `['hypervisors']['facts']['hypervisor.type']`, can't you?
And if you have two of the same, probably you could also filter based on log entry timestamp? If you run your test for VCenter today, obviously VCenter entry from January 2019 is not what we need.

I'm thinking if we should introduce proper parser for this file. It could also parse JSON from entries, so you can drop all of these `re.findall()`. Very rough implementation below:

```python
from pprint import pprint
import json

logs = []

entry = None


def _parse_entry(entry):
    try:
        return json.loads(entry)
    except json.decoder.JSONDecodeError:
        return None
    

with open('rhsm.log') as fh:
    for line in fh:
        if line[0].isdigit():
            if entry:
                logs.append(_parse_entry(entry))
            entry = '{'
            continue
        if entry: 
            entry += line
    else:
        logs.append(_parse_entry(entry))

logs = [_ for _ in logs if _ is not None]
pprint(logs)
```
For my fake log file, it will give something like:
```
[{'hypervisors': [{'facts': {'cpu.cpu_socket(s)': '1',
                             'dmi.system.uuid': '9bff4d56-3b59-6355-ece2-e02094d27d11',
                             'hypervisor.type': 'VMware ESXi',
                             'hypervisor.version': '6.7.0'},
                   'guestIds': [{'attributes': {'active': 1,
                                                'virtWhoType': 'esx'},
                                 'guestId': '564dd2ec-ed5b-658a-359f-2158dab0907f',
                                 'state': 1}],
                   'hypervisorId': {'hypervisorId': 'bootp-73-131-226.rhts.eng.pek2.redhat.com'},
                   'name': 'bootp-73-131-226.rhts.eng.pek2.redhat.com'}]},
 {'hypervisors': [{'facts': {'cpu.cpu_socket(s)': '1',
                             'dmi.system.uuid': '9bff4d56-3b59-6355-ece2-e02094d27d11',
                             'hypervisor.type': 'VMware ESXi',
                             'hypervisor.version': '6.7.0'},
                   'guestIds': [{'attributes': {'active': 1,
                                                'virtWhoType': 'esx'},
                                 'guestId': '666dd2ec-ed5b-658a-359f-2158dab0907f',
                                 'state': 1}],
                   'hypervisorId': {'hypervisorId': 'drugi-bootp-75-131-226.rhts.eng.pek2.redhat.com'},
                   'name': 'drugi-bootp-73-131-226.rhts.eng.pek2.redhat.com'}]}]
```

Do you think this is right direction?

Reply:
@mirzal , I copied the rhsm.log to http://pastebin.test.redhat.com/782546 when virt-who service restart, there are three json sections in this log file because I restarted virt-who service three times. 
I need the following message from the rhsm.log:
1). Should no 'ERROR' message in the log, if error message found, it means virt-who service is not run normally. In this log file, I restarted virt-who service failed in the last time.

2). I need to know the interval time is expected as I set it in the virt-who configure page, the related message is "Starting infinite loop with 3600 seconds interval", it means I set the interval time 1 hour as excepted.

3). I need the `hypervisorId` value from the log file, if virt-who service is running normally and  the hypervisorId is "bootp-73-131-226.rhts.eng.pek2.redhat.com", there should be an entry created and named "virt-who-bootp-73-131-226.rhts.eng.pek2.redhat.com-1" in the Content Hosts page in satellite, and then I can test the entry.

I still think it's necessary to clean the rhsm.log before running virt-who service, because virt-who service will be restarted frequently in our UI/CLI/API testing,  and `isdigit()` is not suitable to get the above message from the log file.

Reply:
I think there is a room to improve our approach, and perhaps also testability of software. But these can take significant time, so let's not block PR on them.

Thanks for this discussion. It's definitely more complex than I initially thought.

Reply:
@mirzal please wait a moment, I have a new commit later